x-wing/src/lib.rs

```diff
@@ -0,0 +1,406 @@
+    pk_x: x25519_dalek::PublicKey,
+}
+
+impl Encapsulate<Ciphertext, SharedSecret> for EncapsulationKey {
+    type Error = ();
+
+    fn encapsulate(
+        &self,
+        rng: &mut impl CryptoRngCore,
+    ) -> Result<(Ciphertext, SharedSecret), Self::Error> {
+        // Swapped order of operations compared to RFC, so that usage of the rng matches the RFC
+        let (ct_m, ss_m) = self.pk_m.encapsulate(rng)?;
+
+        let ek_x: SharedSecret = generate(rng);
+        let ct_x = x25519(ek_x, X25519_BASEPOINT_BYTES);
+        let ss_x = x25519(ek_x, self.pk_x.to_bytes());
+
+        let ss = combiner(&ss_m, &ss_x, &ct_x, &self.pk_x);
+
+        #[cfg(feature = "zeroize")]
+        {
+            let mut ss_x = ss_x;
+            ss_x.zeroize();
+        }
+
+        let ct = Ciphertext { ct_m, ct_x };
+        Ok((ct, ss))
+    }
+}
+
+impl EncapsulationKey {
+    /// Convert the key to the following format:
+    /// ML-KEM-768 public key(1184 bytes) | X25519 public key(32 bytes).
+    #[must_use]
+    pub fn as_bytes(&self) -> [u8; ENCAPSULATION_KEY_SIZE] {
+        let mut buffer = [0u8; ENCAPSULATION_KEY_SIZE];
+        buffer[0..1184].copy_from_slice(&self.pk_m.as_bytes());
+        buffer[1184..1216].copy_from_slice(self.pk_x.as_bytes());
+        buffer
+    }
+}
+
+impl From<&[u8; ENCAPSULATION_KEY_SIZE]> for EncapsulationKey {
+    fn from(value: &[u8; ENCAPSULATION_KEY_SIZE]) -> Self {
+        let mut pk_m = [0; 1184];
+        pk_m.copy_from_slice(&value[0..1184]);
+        let pk_m = MlKem768EncapsulationKey::from_bytes(&pk_m.into());
+
+        let mut pk_x = [0; 32];
+        pk_x.copy_from_slice(&value[1184..]);
+        let pk_x = x25519_dalek::PublicKey::from(pk_x);
+        EncapsulationKey { pk_m, pk_x }
+    }
+}
+
+/// X-Wing decapsulation key or private key.
+#[derive(Clone, PartialEq, Eq)]
+#[cfg_attr(feature = "zeroize", derive(Zeroize, ZeroizeOnDrop))]
+pub struct DecapsulationKey {
+    sk: [u8; DECAPSULATION_KEY_SIZE],
+}
+
+impl Decapsulate<Ciphertext, SharedSecret> for DecapsulationKey {
+    type Error = ();
+
+    fn decapsulate(&self, ct: &Ciphertext) -> Result<SharedSecret, Self::Error> {
+        let (sk_m, sk_x, _pk_m, pk_x) = self.expand_key();
+        let ss_m = sk_m.decapsulate(&ct.ct_m)?;
+        let ss_x = x25519(sk_x.to_bytes(), ct.ct_x);
+        let ss = combiner(&ss_m, &ss_x, &ct.ct_x, &pk_x);
+
+        #[cfg(feature = "zeroize")]
+        {
+            let mut ss_x = ss_x;
+            ss_x.zeroize();
+        }
+
+        Ok(ss)
+    }
+}
+
+impl DecapsulationKey {
+    /// Generate a new `DecapsulationKey` using `OsRng`.
+    #[cfg(feature = "getrandom")]
+    pub fn generate_rng() -> DecapsulationKey {
+        Self::generate(&mut rand_core::OsRng)
+    }
+
+    /// Generate a new `DecapsulationKey` using the provided RNG.
+    pub fn generate(rng: &mut impl CryptoRngCore) -> DecapsulationKey {
+        let sk = generate(rng);
+        DecapsulationKey { sk }
+    }
```

Comment:
Per our [naming conventions](https://github.com/RustCrypto/meta/issues/19) these should be `generate` and `generate_with_rng` respectively.